src/cocaine-app/history.py

```diff
@@ -0,0 +1,401 @@
+    """
+    Represents node backend that was noted as a part of a group at a certain time
+
+    A set of such objects is united under a single GroupNodeBackendsSet object,
+    which takes part in GroupNodeBackendsSetRecord along with record's timestamp and type.
+    """
+    def __init__(self, hostname, port, family, backend_id, path):
+        self.hostname = hostname
+        self.port = port
+        self.family = family
+        self.backend_id = backend_id
+        self.path = path
+
+    def __hash__(self):
+        return hash((self.hostname, self.port, self.family, self.backend_id, self.path))
+
+    def __eq__(self, other):
+        return (self.hostname == other.hostname and
+                self.port == other.port and
+                self.family == other.family and
+                self.backend_id == other.backend_id and
+                self.path == other.path)
+
+    def dump(self):
+        return {
+            'hostname': self.hostname,
+            'port': self.port,
+            'family': self.family,
+            'backend_id': self.backend_id,
+            'path': self.path,
+        }
+
+    def __repr__(self):
+        return (
+            '<GroupNodeBackendRecord: '
+            'hostname={hostname}, '
+            'port={port}, '
+            'family={family}, '
+            'backend_id={backend_id}, '
+            'path={path}>'.format(
+                hostname=self.hostname,
+                port=self.port,
+                family=self.family,
+                backend_id=self.backend_id,
+                path=self.path
+            )
+        )
+
+
+class GroupCoupleRecord(GroupStateRecord):
+    """
+    Record that represents group's affiliation with couple
+
+    Attributes:
+        couple - iterable of integer group ids
+        timestamp - timestamp of record creation
+        type - type of the record, GroupStateRecord.HISTORY_*
+    """
+    def __init__(self, couple, timestamp=None, type=GroupStateRecord.HISTORY_RECORD_AUTOMATIC):
+        super(GroupCoupleRecord, self).__init__(timestamp, type)
+        self.couple = couple
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __eq__(self, other):
+        return set(self.couple) == set(other.couple)
+
+    def dump(self):
+        return {
+            'couple': self.couple,
+            'type': self.type,
+            'timestamp': self.timestamp,
+        }
+
+    def __repr__(self):
+        return '<GroupCoupleRecord: couple {}, timestamp {}, type {}>'.format(
+            self.couple,
+            self.timestamp,
+            self.type
+        )
+
+
+class GroupHistoryNotFoundError(Exception):
+    pass
+
+
+class GroupHistoryFinder(object):
+
+    def __init__(self, db):
+        self.collection = Collection(db[config['metadata']['history']['db']], 'history')
+
+    def _get_group_history(self, group_id):
+        group_history_list = GroupHistory.list(
+            self.collection,
+            group_id=group_id).limit(1)
```

Comment:
I'll try it for some time and think about it. Thanks.
https://github.com/yandex/mastermind/commit/6eeef29e237ee55563a4a1b2481856c209f6a43f